if the server is booting when OnlineManagementClient connects, it should wait until it's running

It's possible for `ManagementClient.online(...)` to end up with an exception like this: `java.lang.AssertionError: Expected success, but operation failed: {"outcome" => "failed","failure-description" => "WFLYCTL0379: System boot is in process; execution of remote management operations is not currently available"}`

This is caused by the client trying to perform a management operation while the server is still booting. The client should detect this situation and wait until the server is running (using some reasonable timeout, e.g. 20 seconds).

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@
 ## 1.0.1 (not yet released)
 
 - updated `GroovyXmlTransform` to understand new XML element `<host-excludes>`
+- added `OnlineOptions.bootTimeout`
 
 ## 1.0.0 [The Ez|Ra Edition](https://en.wikipedia.org/wiki/Embassytown) (2016-03-15)
 

--- a/README.md
+++ b/README.md
@@ -51,7 +51,15 @@ __You have to provide both `*-controller-client` and `*-cli` yourself.__
 If you want to use commands for patching, you also have to provide
 `wildfly-patching`.
 
-For JBoss AS 7:
+Other dependencies might be required if you're using "local" CLI operations,
+such as `jms-queue add ...` on WildFly 10 (it was a built-in operation
+in previous WildFly versions, but was moved to a separate module due to
+messaging changes). These dependencies are not listed here, because there
+are no commands in Creaper that use them.
+
+#### JBoss AS 7
+
+Dependencies for `creaper-core`:
 
     <!-- 7.5.0.Final-redhat-21 is JBoss EAP 6.4.0 -->
     <dependency>
@@ -64,13 +72,18 @@ For JBoss AS 7:
         <artifactId>jboss-as-cli</artifactId>
         <version>7.5.0.Final-redhat-21</version>
     </dependency>
+
+Dependencies for `creaper-commands`:
+
     <dependency>
         <groupId>org.wildfly</groupId>
         <artifactId>wildfly-patching</artifactId>
         <version>7.5.0.Final-redhat-21</version>
     </dependency>
 
-For WildFly 8:
+#### WildFly 8
+
+Dependencies for `creaper-core`:
 
     <dependency>
         <groupId>org.wildfly</groupId>
@@ -82,13 +95,18 @@ For WildFly 8:
         <artifactId>wildfly-cli</artifactId>
         <version>8.2.1.Final</version>
     </dependency>
+
+Dependencies for `creaper-commands`:
+
     <dependency>
         <groupId>org.wildfly</groupId>
         <artifactId>wildfly-patching</artifactId>
         <version>8.2.1.Final</version>
     </dependency>
 
-For WildFly 9, which is based on WildFly Core 1:
+#### WildFly 9 (based on WildFly Core 1)
+
+Dependencies for `creaper-core`:
 
     <dependency>
         <groupId>org.wildfly.core</groupId>
@@ -100,28 +118,44 @@ For WildFly 9, which is based on WildFly Core 1:
         <artifactId>wildfly-cli</artifactId>
         <version>1.0.2.Final</version>
     </dependency>
+
+Dependencies for `creaper-commands`:
+
     <dependency>
         <groupId>org.wildfly.core</groupId>
         <artifactId>wildfly-patching</artifactId>
         <version>1.0.2.Final</version>
     </dependency>
 
-For WildFly 10, which is based on WildFly Core 2:
+#### WildFly 10 (based on WildFly Core 2)
+
+Dependencies for `creaper-core`:
 
     <dependency>
         <groupId>org.wildfly.core</groupId>
         <artifactId>wildfly-controller-client</artifactId>
-        <version>2.0.5.Final</version>
+        <version>2.1.0.Final</version>
     </dependency>
     <dependency>
         <groupId>org.wildfly.core</groupId>
         <artifactId>wildfly-cli</artifactId>
-        <version>2.0.5.Final</version>
+        <version>2.1.0.Final</version>
     </dependency>
+
+    <!-- probably a bug in wildfly-cli dependencies -->
+    <!-- not required if the dependency on wildfly-patching below is also added ->
+    <dependency>
+        <groupId>org.jboss.logmanager</groupId>
+        <artifactId>jboss-logmanager</artifactId>
+        <version>2.0.3.Final</version>
+    </dependency>
+
+Dependencies for `creaper-commands`:
+
     <dependency>
         <groupId>org.wildfly.core</groupId>
         <artifactId>wildfly-patching</artifactId>
-        <version>2.0.5.Final</version>
+        <version>2.1.0.Final</version>
     </dependency>
 
 ### Transitive Dependencies
@@ -140,8 +174,9 @@ on Creaper:
 If you need to bring your own version of some of these libraries, you should
 use dependency exclusions.
 
-Specifically for Groovy, Creaper only needs `groovy` and `groovy-xml`,
-but excluding them and providing `groovy-all` should work too.
+Specifically for Groovy: this is only needed for offline management
+(i.e, XML configuration files manipulation). Creaper only needs `groovy` and
+`groovy-xml`, but excluding them and providing `groovy-all` should work too.
 
 ## Use
 

--- a/core/src/main/java/org/wildfly/extras/creaper/core/ManagementClient.java
+++ b/core/src/main/java/org/wildfly/extras/creaper/core/ManagementClient.java
@@ -26,7 +26,9 @@ public final class ManagementClient {
      * (e.g. there's no server listening on host:port)
      * @throws IllegalStateException if the client is connected to a domain controller even if the {@code options}
      * say that it should be a standalone server, or if the client is connected to a standalone server even if
-     * the {@code options} say that it should be a domain controller
+     * the {@code options} say that it should be a domain controller, or if the client connects to a server
+     * that is still booting and the boot process doesn't finish in time (see
+     * {@link OnlineOptions.OptionalOnlineOptions#bootTimeout(int) OnlineOptions.bootTimeout}).
      */
     public static OnlineManagementClient online(OnlineOptions options) throws IOException {
         return OnlineClientFactory.instance.create(options);

--- a/core/src/main/java/org/wildfly/extras/creaper/core/online/Constants.java
+++ b/core/src/main/java/org/wildfly/extras/creaper/core/online/Constants.java
@@ -44,4 +44,8 @@ public final class Constants extends ClientConstants {
             "JBAS014807", "WFLYCTL0216", // management resource not found
             "JBAS014883", "WFLYCTL0030"  // no resource definition is registered for address
     ));
+
+    public static final List<String> RESULT_CODES_FOR_BOOT_IN_PROGRESS = Collections.unmodifiableList(Arrays.asList(
+            "JBAS013493", "WFLYCTL0379"
+    ));
 }

--- a/core/src/main/java/org/wildfly/extras/creaper/core/online/OnlineOptions.java
+++ b/core/src/main/java/org/wildfly/extras/creaper/core/online/OnlineOptions.java
@@ -1,6 +1,7 @@
 package org.wildfly.extras.creaper.core.online;
 
 import org.jboss.as.controller.client.ModelControllerClient;
+import org.jboss.dmr.ModelNode;
 import org.wildfly.extras.creaper.core.ManagementClient;
 
 import javax.net.ssl.SSLContext;
@@ -15,6 +16,7 @@ import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.util.Collections;
 import java.util.Map;
+import java.util.concurrent.TimeoutException;
 
 /**
  * This basically follows the builder pattern, but ensures on the type level that everything that must be set is
@@ -35,6 +37,7 @@ public final class OnlineOptions {
     final int port;
     private final ManagementProtocol protocol; // this can be "null" for unspecified protocol
     private final int connectionTimeout;
+    private final int bootTimeout;
 
     private final String username;
     private final String password;
@@ -64,6 +67,7 @@ public final class OnlineOptions {
         this.port = data.port;
         this.protocol = data.protocol;
         this.connectionTimeout = data.connectionTimeout;
+        this.bootTimeout = data.bootTimeout;
         this.username = data.username;
         this.password = data.password;
         this.localAuthDisabled = data.localAuthDisabled;
@@ -88,6 +92,7 @@ public final class OnlineOptions {
 
         private ManagementProtocol protocol; // often not set at all
         private int connectionTimeout;
+        private int bootTimeout = 10000;
 
         private String username;
         private String password;
@@ -252,13 +257,29 @@ public final class OnlineOptions {
             return this;
         }
 
-        /** Timeout to enforce when connecting to the server. Optional, must be {@code > 0}. */
-        public OptionalOnlineOptions connectionTimeout(int connectionTimeout) {
-            if (connectionTimeout <= 0) {
+        /**
+         * Timeout to use when connecting to the server. Optional, must be {@code > 0}.
+         * By default, no timeout is used.
+         */
+        public OptionalOnlineOptions connectionTimeout(int timeoutInMillis) {
+            if (timeoutInMillis <= 0) {
                 throw new IllegalArgumentException("Connection timeout must be > 0");
             }
 
-            data.connectionTimeout = connectionTimeout;
+            data.connectionTimeout = timeoutInMillis;
+            return this;
+        }
+
+        /**
+         * Timeout to use when waiting for the server to boot. Optional. A value {@code <= 0} means "no timeout".
+         * By default, a value of {@code 10000} is used (i.e., a 10 seconds timeout).
+         */
+        public OptionalOnlineOptions bootTimeout(int timeoutInMillis) {
+            if (timeoutInMillis <= 0) {
+                timeoutInMillis = 0;
+            }
+
+            data.bootTimeout = timeoutInMillis;
             return this;
         }
 
@@ -336,6 +357,8 @@ public final class OnlineOptions {
             saslOptions = Collections.singletonMap("SASL_DISALLOWED_MECHANISMS", "JBOSS-LOCAL-USER");
         }
 
+        ModelControllerClient modelControllerClient;
+
         // the variant with the "protocol" parameter exists since WildFly 8, and if it is available, it is preferred
         // to the protocol-less variant available in JBoss AS 7, because we want to choose the protocol dynamically
         //
@@ -357,7 +380,7 @@ public final class OnlineOptions {
                 protocolName = protocol.protocolName();
             }
 
-            return (ModelControllerClient) createMethod.invoke(null, // static method
+            modelControllerClient = (ModelControllerClient) createMethod.invoke(null, // static method
                     protocolName, host, port, callbackHandler, null, connectionTimeout, saslOptions);
         } catch (NoSuchMethodException e) {
             if (protocol == ManagementProtocol.HTTP_REMOTING) {
@@ -366,8 +389,8 @@ public final class OnlineOptions {
                         + CREAPER_WILDFLY + "' system property was set), but client libraries are AS7-only");
             }
 
-            return ModelControllerClient.Factory.create(host, port, callbackHandler, null, connectionTimeout,
-                    saslOptions);
+            modelControllerClient = ModelControllerClient.Factory.create(host, port,
+                    callbackHandler, null, connectionTimeout, saslOptions);
         } catch (InvocationTargetException e) {
             if (e.getCause() instanceof IOException) {
                 throw (IOException) e.getCause();
@@ -375,6 +398,60 @@ public final class OnlineOptions {
             throw new IOException(e);
         } catch (IllegalAccessException e) {
             throw new IOException(e);
+        }
+
+        try {
+            connectAndWaitUntilServerBoots(modelControllerClient, bootTimeout);
+        } catch (Exception e) {
+            modelControllerClient.close();
+
+            if (e instanceof IOException) {
+                throw (IOException) e;
+            } else if (e instanceof IllegalStateException) {
+                throw (IllegalStateException) e;
+            } else {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        return modelControllerClient;
+    }
+
+    private static void connectAndWaitUntilServerBoots(ModelControllerClient client, int timeoutInMillis)
+            throws IOException, InterruptedException, TimeoutException {
+        ModelNode op = new ModelNode();
+        op.get(Constants.OP).set(Constants.WHOAMI);
+        op.get(Constants.OP_ADDR).setEmptyList();
+
+        long endTime = System.currentTimeMillis() + timeoutInMillis;
+        while (System.currentTimeMillis() < endTime) {
+            ModelNodeResult result = new ModelNodeResult(client.execute(op));
+
+            if (result.isSuccess()) {
+                return;
+            }
+
+            boolean stillBooting = false;
+
+            String failureDescription = result.get(Constants.FAILURE_DESCRIPTION).asString();
+            for (String code : Constants.RESULT_CODES_FOR_BOOT_IN_PROGRESS) {
+                if (failureDescription.startsWith(code)) {
+                    stillBooting = true;
+                    break;
+                }
+            }
+
+            if (stillBooting) {
+                Thread.sleep(100);
+            } else {
+                // shouldn't happen
+                throw new IllegalStateException("Unknown server state: " + failureDescription);
+            }
+        }
+
+        ModelNodeResult result = new ModelNodeResult(client.execute(op));
+        if (!result.isSuccess()) {
+            throw new TimeoutException("Waiting for server to boot timed out");
         }
     }
 

